### ISSUE TITLE
FDS 6.6.0 - DAPL startup: RLIMIT_MEMLOCK too small

### ISSUE THREAD

**Issue by gsztar** [firemodels/fds#5670]
I run my case on FDS 6.6.0 and I received following warnings:
fds@n1:~/fds/n1/SD1_test6_6$ mpiexec -np 8 -hosts n1,n2  fds SD1_test.fds
**_[0] DAPL startup: RLIMIT_MEMLOCK too small
[1] DAPL startup: RLIMIT_MEMLOCK too small
[2] DAPL startup: RLIMIT_MEMLOCK too small
[3] DAPL startup: RLIMIT_MEMLOCK too small
[4] DAPL startup: RLIMIT_MEMLOCK too small
[5] DAPL startup: RLIMIT_MEMLOCK too small
[6] DAPL startup: RLIMIT_MEMLOCK too small
[7] DAPL startup: RLIMIT_MEMLOCK too smal_**l
 Mesh 1 is assigned to MPI Process 0
 Mesh 2 is assigned to MPI Process 1
 Mesh 3 is assigned to MPI Process 1
 Mesh 4 is assigned to MPI Process 1
 Mesh 5 is assigned to MPI Process 1
 Mesh 6 is assigned to MPI Process 1
 Mesh 7 is assigned to MPI Process 2
 Mesh 8 is assigned to MPI Process 2
 Mesh 9 is assigned to MPI Process 3
 Mesh 10 is assigned to MPI Process 4
 Mesh 11 is assigned to MPI Process 5
 Mesh 12 is assigned to MPI Process 6
 Mesh 13 is assigned to MPI Process 6
 Mesh 14 is assigned to MPI Process 7
 Completed Initialization Step  1
 Completed Initialization Step  2
 Completed Initialization Step  3
 Completed Initialization Step  4

 Fire Dynamics Simulator

 Current Date     : November  3, 2017  15:08:08
 Version          : FDS 6.6.0
 Revision         : FDS6.6.0-131-g88ae75a-HEAD
 Revision Date    : Wed Nov 1 16:03:29 2017 -0400
 Compiler         : Intel ifort 17.0.4
 Compilation Date : Nov 01, 2017 21:03:34

 MPI Enabled;    Number of MPI Processes:       8
 OpenMP Enabled; Number of OpenMP Threads:      1

 MPI version: 3.1
 MPI library version: Intel(R) MPI Library 2017 Update 3 for Linux* OS

 Job TITLE        :
 Job ID string    : SD1_test

 Time Step:      1, Simulation Time:      0.17 s
 Time Step:      2, Simulation Time:      0.33 s
 Time Step:      3, Simulation Time:      0.50 s
 Time Step:      4, Simulation Time:      0.67 s
 Time Step:      5, Simulation Time:      0.83 s

How I can solve that problem? I use Ubuntu 15.04 (GNU/Linux 3.19.0-58-generic x86_64).

**Comment by sbenkorichi** [comment]
Thanks,
We'll take a look.

**Comment by sbenkorichi** [comment]
Do me this:
`$ ulimit -a `
Share what it gives.

**Comment by gsztar** [comment]
fds@n1:~/fds/n1/SD1_test6_6$ ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 63613
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 63613
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited

**Comment by sbenkorichi** [comment]
Ok,
I think I now understand your issue. 
On terminal do:
`$ ulimit -l unlimited `
On the same terminal run the case (must be on the same). If it works, then you need to make this change permanent, i.e. you have to put it either in your .bashrc file or /etc/security/limits.conf . But, let's first check if that solves it or not, and then we can decide where you put it.

**Comment by gsztar** [comment]
fds@n1:~/fds/n1/SD1_test6_6$ ulimit -l unlimited
-bash: ulimit: max locked memory: cannot modify limit: Operation not permitted
fds@n1:~/fds/n1/SD1_test6_6$ mpiexec -np 8 -hosts n1,n2  fds SD1_test.fds[0] DAPL startup: RLIMIT_MEMLOCK too small
[4] DAPL startup: RLIMIT_MEMLOCK too small
[1] DAPL startup: RLIMIT_MEMLOCK too small
[5] DAPL startup: RLIMIT_MEMLOCK too small
[2] DAPL startup: RLIMIT_MEMLOCK too small
[6] DAPL startup: RLIMIT_MEMLOCK too small
[3] DAPL startup: RLIMIT_MEMLOCK too small
[7] DAPL startup: RLIMIT_MEMLOCK too small
 Mesh 1 is assigned to MPI Process 0
 Mesh 2 is assigned to MPI Process 1
 Mesh 3 is assigned to MPI Process 1
 Mesh 4 is assigned to MPI Process 1
 Mesh 5 is assigned to MPI Process 1
 Mesh 6 is assigned to MPI Process 1
 Mesh 7 is assigned to MPI Process 2
 Mesh 8 is assigned to MPI Process 2
 Mesh 9 is assigned to MPI Process 3
 Mesh 10 is assigned to MPI Process 4
 Mesh 11 is assigned to MPI Process 5
 Mesh 12 is assigned to MPI Process 6
 Mesh 13 is assigned to MPI Process 6
 Mesh 14 is assigned to MPI Process 7
 Completed Initialization Step  1
 Completed Initialization Step  2
 Completed Initialization Step  3
 Completed Initialization Step  4

**Comment by marcosvanella** [comment]
You will need to do this as superuser:
$ sudo ulimit -l unlimited

**Comment by sbenkorichi** [comment]
Well, you need to get ownernship.
try it with : 
$ sudo ulimit -l unlimited 
Check if it's set to unlimited by : 
$ ulimit -l 
Does it give unlimited or 64?
If it doesn't work, then let's try this second way:

**Comment by gsztar** [comment]
yes, it does 64.
And doesn't work. What is second way?

**Comment by sbenkorichi** [comment]
Second way:

$ sudo vi /etc/security/limits.conf

Add this line at the end of file to set it to unlimited 


```
*          hard   memlock           unlimited
```

Reboot the system. Login and verify that the locked memory is now set to unlimited:

$ ulimit –l
unlimited

**Comment by marcosvanella** [comment]
Also, do this in both nodes.

**Comment by gsztar** [comment]
After second way:
_fds@n1:~$ ulimit -l
64_

**Comment by gsztar** [comment]
![etc](https://user-images.githubusercontent.com/16048248/32384296-bac3bd14-c0ba-11e7-810c-ee809ef0ef00.JPG)

**Comment by sbenkorichi** [comment]
Alright,
Before, that let's try this: 
Looks like your memeory size also isn't set to unlimited:
`$ sudo ulimit -s unlimited `
check 
`$ ulimit -s  `
What does it give?

**Comment by gsztar** [comment]
8192

**Comment by sbenkorichi** [comment]
Weird, 
You don't seem to have permission.
Try this: 
`$ sudo sysctl -w fs.inotify.max_user_watches=100000`
`$ ulimit -l`

**Comment by gsztar** [comment]
fds@n1:~$ sudo sysctl -w fs.inotify.max_user_watches=100000
fs.inotify.max_user_watches = 100000
fds@n1:~$ ulimit -l
64

**Comment by gsztar** [comment]
fds@n1:~$ sudo ulimit -s unlimited
sudo: ulimit: command not found

**Comment by sbenkorichi** [comment]
Try to do it as a root.
Do this: 
```
$ sudo -i 
$ ulimit -l unlimited 
$ ulimit -l

```

**Comment by gsztar** [comment]
Yes !! Success !!
root@n1:~# ulimit -l unlimited
root@n1:~# ulimit -l
unlimited
root@n1:~#

**Comment by sbenkorichi** [comment]
As a root, in the same directory of your inputfile: Make sure ulimit -l is set to unlimited, if so then run fds and see what it gives.

**Comment by gsztar** [comment]
How I can go directory of inputfile as root?

**Comment by sbenkorichi** [comment]
Here is an example: 
```
sbenkorichi ~ # 
sbenkorichi ~ # cd /home/salah/
```
Then you can cd to your input file directory.

**Comment by gsztar** [comment]
root@n1:/home/fds/fds/n1/SD1_test6_6# mpiexec -np 8 -hosts n1,n2 fds SD1_test.fds
The program 'mpiexec' can be found in the following packages:
 * lam-runtime
 * mpich
 * openmpi-bin
Try: apt-get install <selected package>

**Comment by sbenkorichi** [comment]
The root isn't seeing mpiexec, even by providing the full path, I'm not sure it would help as it might need access to certain libraries. 
Let me think of a way how we can overcome the issue of the memory lock.

**Comment by sbenkorichi** [comment]
Ok, 
I think I found your issue, put these two lines in your limits.conf file 
```
*          hard   memlock           unlimited
*          soft    memlock           unlimited
```
DO NOT PUT` #` before them. That will disable them.

2. Do this: 
`$ su USER --shell /bin/bash --command "ulimit -n"`
change USER to your username, if you don't know it do : `$ whoami `to know your username.
It should return : unlimited 
Then you need to reboot so the changes can take affect. 
3. test it: `$ ulimit -l`


Let's see how it goes.

**Comment by gsztar** [comment]
fds@n1:~$ su fds --shell /bin/bash --command "ulimit -n"
It returned: 1024

What will be when instead * I put USER?

**Comment by sbenkorichi** [comment]
Sorry, not n, n is for files. Change it please to -l .  `$ su USER --shell /bin/bash --command "ulimit -l"`
* replaces USER, i.e. works for all users on that machine. This just to avoid certain conflicts.
You can put your username instead of *, but I'm afraid that you might have trouble with the other computer. You can experiment and see what it gives.

**Comment by gsztar** [comment]
return : 64

**Comment by sbenkorichi** [comment]
Please reboot. It wouldn't work if you don't reboot.

**Comment by gsztar** [comment]
Works! Returned: unlimited

**Comment by sbenkorichi** [comment]
Good, now try to run fds and see what if it has fixed the warning.

**Comment by gsztar** [comment]
Salah,
you are great! Works! To confirm:

fds@n1:~/fds/n1/SD1_test6_6$ mpiexec -np 8 -hosts n1,n2 fds SD1_test.fds
 Mesh 1 is assigned to MPI Process 0
 Mesh 2 is assigned to MPI Process 1
 Mesh 3 is assigned to MPI Process 1
 Mesh 4 is assigned to MPI Process 1
 Mesh 5 is assigned to MPI Process 1
 Mesh 6 is assigned to MPI Process 1
 Mesh 7 is assigned to MPI Process 2
 Mesh 8 is assigned to MPI Process 2
 Mesh 9 is assigned to MPI Process 3
 Mesh 10 is assigned to MPI Process 4
 Mesh 11 is assigned to MPI Process 5
 Mesh 12 is assigned to MPI Process 6
 Mesh 13 is assigned to MPI Process 6
 Mesh 14 is assigned to MPI Process 7
 Completed Initialization Step  1
 Completed Initialization Step  2
 Completed Initialization Step  3
 Completed Initialization Step  4

**Comment by sbenkorichi** [comment]
Happy days !
But first let's work also your stack memory, otherwise it will hit you next time. 
try to add this to your .bashrc file:
` ulimit -s unlimited `
Log off and log in again. Check if it's working as we want it, it should return unlimited: 
`$ ulimit -s `
Let's see how it goes this time.

**Comment by gsztar** [comment]
Just: ulimit -s unlimited ?

**Comment by sbenkorichi** [comment]
Yes, Just add that line into your .bashrc startup so each time you login, it will be set by default.

**Comment by gsztar** [comment]
Works !!! Returned _unlimited_

**Comment by sbenkorichi** [comment]
Good, then this issue has been fixed. You're good to use FDS now.

**Comment by gsztar** [comment]
Last question for today :)
What should the hostfile include?

**Comment by sbenkorichi** [comment]
Oh sorry, I forgot about that. I will leave @mcgratta or @rmcdermo  as they have a cluster. 
What do you recommend as a setting for hostfile when there are two computers each has 4 cores. 
My understanding is that you set it as : 
comp1 slots=4
comp2 slots=4

Is that what you recommend?

**Comment by gsztar** [comment]
I just wrote: mpiexec -np 8 -hosts comp1, comp2 fds file.fds.
I am not sure that: 
comp1 slots=4
comp2 slots=4
are correct.
@mcgratta Could you please help?

**Comment by mcgratta** [comment]
Try this
```
mpiexec -hosts 2 comp1 4 comp2 4 fds fds.fds
```
This command does not use a hostfile. While the job is running, login to each node and type
```
top
```
Then hit 1 to see how the CPUs are dividing the workload. You want to see each core that you use working 100%.

**Comment by gsztar** [comment]
It is works. Thank you. Special greetings to @sbenkorichi.

**Comment by sbenkorichi** [comment]
Ok, I was able to building a small cluster, I've added another 4 cores to my original 8 cores, now I've got 12 processus running.
I've received similar warning, setting up the limits.conf file get rid of the warnings. 
As for running the case. I set a dummy file called machinefile , it has the following:
```
sbenkorichi=8  # <=== My original machine (setting it as master)
sbenkorichi2=4 #<===My new added node (slave node)
```
I run the case as the following: 
My dummy test file has 12 meshes, so:
`$ mpiexec -np 12 -f machinefile fds test.fds `
Checking with top, everything looks good.
